### ISSUE TITLE
[4.0.0-rc.5] NavigationEnd does not update Title/Meta server-side

### ISSUE THREAD

**Issue by nicky-lenaers** [angular/angular#15368]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
When using NavigationEnd event, the `<title>` and `<meta>` tags are not updated server-side.

**Expected behavior**
Update `<title>` and `<meta>` server-side.

**Minimal reproduction of the problem with instructions**
Say I have a route with custom data:
```sh
...
{
    path: 'home',
    loadChildren: './+home/home.module#HomeModule',
    data: {
        title: 'Home Title'
    }
}
...
```

I used [this](https://toddmotto.com/dynamic-page-titles-angular-2-router-events) implementation by Todd Moto to update the page title on route change, like so:

```sh
import { Title } from '@angular/platform-browser';
import { ActivatedRoute, Router, NavigationStart, NavigationEnd } from '@angular/router';
...
export class AppComponent implements OnInit {

    constructor(
        private titleService: Title,
	private router: Router,
	private activatedRoute: ActivatedRoute) { }
    ...
    ngOnInit() {
        this.router.events
            .filter(event => event instanceof NavigationEnd)
            .map(() => this.activatedRoute)
            .map(route => {
                while (route.firstChild) route = route.firstChild;
                return route;
            })
            .filter(route => route.outlet === 'primary')
            .mergeMap(route => route.data)
            .subscribe((event) => this.titleService.setTitle(event['title']));
      }
...
}
```
Using server-side rendering by using `@angular/platform-server`, if I inspect the page source after route change, I see that the `<title>` is not updated with the new value. It **does** display it in the browser tab though.

When I use the `NavigationStart` event instead of `NavigationEnd` and replace `setTitle()` as follows...
```sh
this.titleService.setTitle('Home Title')); // hard-coded value, non-dynamic
```
... it actually updated the value if I inspect the page source (server-side rendered).

But with the `NavigationStart` event, the data in `ActivatedRoute` (`this.activatedRoute`) reflects the route **before** change, not the route the user is about to go to. Thus, I cannot reach the data of the next route to update the page title server-side.

Above example holds for `<meta>` tags as well.

**What is the motivation / use case for changing the behavior?**
SEO of page title / meta

**Please tell us about your environment:**

* **Angular version:**
4.0.0-rc.5

* **Browser:** 
all
 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
TypeScript

* **Node (for AoT issues):** 
`node --version` =   7.7.0

**Comment by Toxicable** [comment]
Would you mind providing a plunkr reproducing this issue

**Comment by Toxicable** [comment]
Could you please provide a minimal reproduction in a repo that we can see the issue in?

**Comment by nicky-lenaers** [comment]
@Toxicable will do in a short while

**Comment by vikerman** [comment]
A repro would be good. Are you using renderModuleFactory to get the page source on the server?

I am guessing there is a race condition between NavigationEnd and applicationRef.isStable() . Will try to see if there is a different way you could get the currently navigated route without having to wait for NavigationEnd.

**Comment by nicky-lenaers** [comment]
It appears to be working on a fresh fork of [ng-universal-demo](https://github.com/FrozenPandaz/ng-universal-demo). I'm still debugging my own app and trying to figure out why it's not working properly.

**Comment by nicky-lenaers** [comment]
@Toxicable ok, found it. Turns out there was another error on the server on each request I made, where I use `this._document.getElementById('main-content').scrollTop = 0;`. I inspected it and it seemed to throw an error on the server (d0h, no document).

I'll have to update that and use `isPlatformBrowser(PLATFORM_ID)` to encapsulate that statement. Once I commented out that line, no errors occurred server-side and it all works well.

Apologies for the confusion, it seems this works well for Universal!

Thanks for the responses!

Closing this issue.
